src/odemis/acq/orsay_milling.py

```diff
@@ -0,0 +1,243 @@
+            if rect[0] == rect[2]:
+                logging.debug("Milling a vertical line. Orsay doesn't like rectangles with 0 "
+                              "height. Line height automatically set equal to the probe size to fit a single line.")
+                obj_to_mill.primaryAxisX = probe_size / 2
+                obj_to_mill.secondaryAxisLength = (rect[3] - rect[1]) / 2 * fov
+            elif rect[1] == rect[3]:
+                logging.debug("Milling a horizontal line. Orsay doesn't like rectangles with 0 "
+                              "width. Line width automatically set equal to the probe size to fit a single line.")
+                obj_to_mill.primaryAxisX = (rect[2] - rect[0]) / 2 * fov
+                obj_to_mill.secondaryAxisLength = probe_size / 2
+            else:
+                obj_to_mill.primaryAxisX = (rect[2] - rect[0]) / 2 * fov
+                obj_to_mill.secondaryAxisLength = (rect[3] - rect[1]) / 2 * fov
+        # common attribute value for point and rectangle, hence outside of if block
+        obj_to_mill.NumberOfPasses = iteration
+        obj_to_mill.totalScanTime = duration * iteration
+
+        obj_to_mill.centerOfMassX = ((rect[2] + rect[0]) / 2 - 0.5) * fov
+        obj_to_mill.centerOfMassY = (0.5 - (rect[3] + rect[1]) / 2) * fov
+
+        lay = Layer()
+        lay.probeSize = probe_size
+        lay.AddForm(0, obj_to_mill)
+
+        proc = Procedure()
+        proc.AddLayer(0, lay)
+        proc.name = f"Odemis rectangle milling {time.time()}"
+
+        # this is going the start the milling indirectly
+        miller.ProcedureInfoSerialNumber.Subscribe(self.on_procedure_info)
+
+        # send the procedure to the server.
+        proc_string = proc.ToString()
+        miller.SetCurrentProcedure(proc_string)
+        logging.debug("Sending procedure with name %s, using FoV = %s m", proc_string, fov)
+
+        milling_activation_timeout = time.time() + 10
+        # wait for milling action to start
+        while miller.MillingActivationState.Actual != '1':
+            time.sleep(0.1)
+            if time.time() > milling_activation_timeout:
+                raise TimeoutError("Milling failed to start up after trying for 10 seconds.")
+
+        logging.debug("Now milling a rectangle %sm by %sm during %s passes of %s seconds",
+                      obj_to_mill.primaryAxisX*2, obj_to_mill.secondaryAxisLength*2, iteration, duration)
+
+        milling_execution_timeout = time.time() + duration * iteration + 5
+        while miller.MillingActivationState.Actual != '0':
+            time.sleep(0.1)
+            if time.time() > milling_execution_timeout:
+                raise TimeoutError("Milling operation timed out. Please check whether the milling is done correctly.")
+        if self.future.cancelled():
+            raise futures.CancelledError()
+
+        logging.debug("Milling completed.")
+
+        miller.MillingActivationState.Unsubscribe(self.on_milling_state)
```

Comment:
That line must *always* be called after the end of the function, even if an exception happens.
=> place that line in a `finally` block, with a `try:` starting after line 135.